src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/common/actions/BaseResourceAllocationAction.java

```diff
@@ -143,6 +146,15 @@ protected abstract void allocate(T stateMachine)
      */
     protected abstract void onFailure(T stateMachine);
 
+
+    /**
+     * Called in a case of allocation failure.
+     */
+    protected void recordFailureReason(T stateMachine, String reason) {
```

Comment:
As for me - the abstract method is better. It will not allow forgetting to define it in new code.